toolchain/parse/handle_paren_expr.cpp

```diff
@@ -6,6 +6,38 @@
 
 namespace Carbon::Parse {
 
+auto HandleOnlyParenExpr(Context& context) -> void {
+  auto state = context.PopState();
+
+  // Advance past the open paren.
+  context.AddLeafNode(NodeKind::ExprOpenParen,
+                      context.ConsumeChecked(Lex::TokenKind::OpenParen));
+
+  context.PushState(state, State::ParenExprFinish);
+  context.PushState(State::OnlyParenExprFinish);
+  context.PushState(State::Expr);
+}
+
+auto HandleOnlyParenExprFinish(Context& context) -> void {
+  auto state = context.PopState();
+
+  if (!context.PositionIs(Lex::TokenKind::CloseParen)) {
+    if (!state.has_error) {
+      CARBON_DIAGNOSTIC(UnexpectedTokenInCompoundMemberAccess, Error,
+                        "Expected `)`.");
+      context.emitter().Emit(*context.position(),
+                             UnexpectedTokenInCompoundMemberAccess);
+      context.ReturnErrorOnState();
+    }
+
+    // Recover from the invalid token.
+    auto end_of_element = context.FindNextOf({Lex::TokenKind::CloseParen});
```

Comment:
Please add a test for if there are multiple close parens and the first one is the wrong one, such as: `.((()))`

See Context::SkipMatchingGroup() for an example of how to handle this, I think you want something like `context.SkipTo(context.tokens().GetMatchedClosingToken(state.token));`

Reply:
Hmm, maybe that's not the best example for a bug. Perhaps something like: `.(x x x ())`

Reply:
Done. The differences are not visible since the output doesn't indicate which `)` token is being referenced.

Reply:
Yeah, I was expecting it to only be visible in the old path because I'd expect processing to resume at the second `)`, yielding a parse error because `)` shouldn't be encountered like that. Skipping to it directly means that parse error is avoided.